### ISSUE TITLE
Verify that addresses reported by identify are correct before inserting them in the DHT

### ISSUE THREAD

**Issue by tomaka** [paritytech/substrate#4920]
Right now nodes tell us the addresses they are reachable from, and we immediately insert them in the DHT.
Instead, we should first attempt to connect to these addresses to see if they're reachable.

On the implementation side, the good news is that Kademlia should automatically detect us reaching the node and add the address we connected to to the DHT, without having to call any method (although I'm not sure that this is the case, but I think that it should be the right logic).

In other words, all we have to do is call `Swarm::dial` for each address that a node reports through identity.

However, this is complicated by the "one connection per node" policy. If we connect to a node we're already connected to, we're going to drop the existing connection.
In other words, we first have to land https://github.com/libp2p/rust-libp2p/pull/1440

cc @romanb @arkpar

**Comment by twittner** [comment]
We have to be careful though not to introduce a DDOS vector by just dialling every reported address. Rate-limiting the dialling attempts may help a bit.

**Comment by tomaka** [comment]
Even after #5099 there is still a problem to solve:

At the moment we insert nodes in the DHT as soon as we connect to them. We should instead only insert in the DHT nodes that belong to the correct chain.

In other words, we should wait until we know that node is on chain A before inserting it in the DHT of chain A.
